deployment/params.py

```diff
@@ -248,34 +309,184 @@ class ConstructorParameters:
+                    check_for_proxy_instances=False,
+                ),
+            )
+            contracts_proxy_info.update({contract_name: proxy_info})
+
+        return cls(contracts_proxy_info=contracts_proxy_info)
+
+    def contract_needs_proxy(self, contract_name) -> bool:
+        proxy_info = self.contracts_proxy_info.get(contract_name)
+        return proxy_info is not None
+
+    def resolve(self, contract_name: str) -> typing.Tuple[ContractContainer, OrderedDict]:
+        """
+        Resolves the proxy data for a single contract.
+        """
+        proxy_info = self.contracts_proxy_info.get(contract_name)
+        if not proxy_info:
+            raise ValueError(f"Unexpected contract to proxy: {contract_name}")
+
+        contract_container = proxy_info.contract_type_container
+
+        resolved_params = _resolve_params(parameters=proxy_info.constructor_params)
+        return contract_container, resolved_params
+
+    @classmethod
+    def _generate_proxy_info(cls, contract_data, variable_context: VariableContext) -> ProxyInfo:
+        proxy_data = contract_data[CONTRACT_PROXY_PARAMETER_KEY] or dict()
+
+        contract_type = variable_context.contract_name
+        if cls.CONTRACT_TYPE in proxy_data:
+            contract_type = proxy_data[cls.CONTRACT_TYPE]
+        contract_type_container = get_contract_container(contract_type)
+
+        constructor_data = cls._default_proxy_parameters(variable_context.contract_name)
+        if CONTRACT_CONSTRUCTOR_PARAMETER_KEY in proxy_data:
+            proxy_constructor_params = proxy_data[CONTRACT_CONSTRUCTOR_PARAMETER_KEY]
+            if "_logic" in proxy_constructor_params:
+                raise cls.Invalid(
+                    "'_logic' parameter cannot be specified: it is implicitly "
+                    "the contract being proxied"
+                )
+
+            constructor_data.update(proxy_constructor_params)
+
+        processed_values = _process_raw_values(constructor_data, variable_context)
+        proxy_info = cls.ProxyInfo(
+            contract_type_container=contract_type_container, constructor_params=processed_values
+        )
+        return proxy_info
+
+    @classmethod
+    def _default_proxy_parameters(cls, contract_name: str) -> OrderedDict:
+        default_parameters = OrderedDict(
+            {"_logic": f"${contract_name}", "initialOwner": "$deployer", "_data": b""}
+        )
+        return default_parameters
+
+
+def _validate_transaction_args(
+    method: ContractTransactionHandler, args: typing.Tuple[Any, ...]
+) -> typing.Dict[str, Any]:
+    """Validates the transaction arguments against the function ABI."""
+    expected_length_abis = [abi for abi in method.abis if len(abi.inputs) == len(args)]
+    for abi in expected_length_abis:
+        named_args = {}
+        for arg, abi_input in zip(args, abi.inputs):
+            if not w3.is_encodable(abi_input.type, arg):
+                break
```

Comment:
Perhaps we should throw here instead? Or is there a scenario where we knowingly take in a non-encodable param and we just want to skip it during this check?

Reply:
Good question. 

Solidity supports function overloading, so there can be multiple methods with the same name, but the abi is different in terms of parameter length or parameter types. We initially filter out the different length cases, but then need to iterate through all resulting abis checking types. The first type mismatch we get, we don't want to raise because subsequent checks for other versions of the method may work.

Only if we make it out of the outer loop without ever returning the `else` block of the inner loop, then we know that there are no matching method abis, and can raise the exception.